cura/Machines/MaterialManager.py

```diff
@@ -308,73 +114,26 @@ def getMaterialGroupListByGUID(self, guid: str) -> Optional[List[MaterialGroup]]
-
-        nodes_to_check = [buildplate_node, nozzle_node, machine_node, default_machine_node]
-        # Fallback mechanism of finding materials:
-        #  1. buildplate-specific material
-        #  2. nozzle-specific material
-        #  3. machine-specific material
-        #  4. generic material (for fdmprinter)
-        machine_exclude_materials = machine_definition.getMetaDataEntry("exclude_materials", [])
-
-        material_id_metadata_dict = dict()  # type: Dict[str, MaterialNode]
-        excluded_materials = set()
-        for current_node in nodes_to_check:
-            if current_node is None:
-                continue
-
-            # Only exclude the materials that are explicitly specified in the "exclude_materials" field.
-            # Do not exclude other materials that are of the same type.
-            for material_id, node in current_node.material_map.items():
-                if material_id in machine_exclude_materials:
-                    excluded_materials.add(material_id)
-                    continue
-
-                if material_id not in material_id_metadata_dict:
-                    material_id_metadata_dict[material_id] = node
-
-        if excluded_materials:
-            Logger.log("d", "Exclude materials {excluded_materials} for machine {machine_definition_id}".format(excluded_materials = ", ".join(excluded_materials), machine_definition_id = machine_definition_id))
-
-        return material_id_metadata_dict
+    ##  Gives a dictionary of all root material IDs and their associated
+    #   MaterialNodes from the ContainerTree that are available for the given
+    #   printer and variant.
+    def getAvailableMaterials(self, definition_id: str, nozzle_name: Optional[str]) -> Dict[str, MaterialNode]:
+        return ContainerTree.getInstance().machines[definition_id].variants[nozzle_name].materials
 
     #
     # A convenience function to get available materials for the given machine with the extruder position.
     #
     def getAvailableMaterialsForMachineExtruder(self, machine: "GlobalStack",
-                                                extruder_stack: "ExtruderStack") -> Optional[Dict[str, MaterialNode]]:
-        buildplate_name = machine.getBuildplateName()
+                                                extruder_stack: "ExtruderStack") -> Dict[str, MaterialNode]:
         nozzle_name = None
         if extruder_stack.variant.getId() != "empty_variant":
             nozzle_name = extruder_stack.variant.getName()
-        diameter = extruder_stack.getApproximateMaterialDiameter()
 
         # Fetch the available materials (ContainerNode) for the current active machine and extruder setup.
-        return self.getAvailableMaterials(machine.definition, nozzle_name, buildplate_name, diameter)
+        materials = self.getAvailableMaterials(machine.definition.getId(), nozzle_name)
+        compatible_material_diameter = str(round(extruder_stack.getCompatibleMaterialDiameter()))
```

Comment:
Maybe still use `ExtruderStack.getApproximateMaterialDiameter()` here? so you don't need to use `round()` to calculate it every time.

`str(extruder_stack.getApproximateMaterialDiameter())`

Another thing is the variable name `compatible_material_diameter`. Later in the function `getDefaultMaterial()`, such as variable is named as `approximate_material_diameter`. I think it should be consistent here.

Those diameters are difficult to name. Here's some clarification:
 - `getCompatibleMaterialDiameter()` returns the material diameter this extruder is designed for, such as 2.85. This is an exact number.
 - `getApproximateMaterialDiameter()` returns the `round(getCompatibleMaterialDiameter())`, which is mainly used to decide which materials should be shown for a certain machine.

Reply:
Done: 614c2970ea6a313ed45949d613d7f49fad1a4cc0